Cabal/Distribution/Simple/Configure.hs

```diff
@@ -570,19 +570,31 @@ configure (pkg_descr0', pbi) cfg = do
       die $ "Cannot build some foreign libraries: "
          ++ intercalate "," unsupportedFLibs
 
-    -- Configure known/required programs & external build tools.
-    -- Exclude build-tool deps on "internal" exes in the same package
+    -- Configure certain external build tools, see below for which ones.
     --
-    -- TODO: Factor this into a helper package.
-    ---
     -- TODO: HACK we're currently ignoring wildcard deps as we don't know
     -- precisely which executables the resolved package will contain.
-    let requiredBuildTools =
-          [ LegacyExeDependency (unUnqualComponentName eName) versionRange
-          | bi <- enabledBuildInfos pkg_descr enabled
-          , buildTool@(ExeDependency _ (Just eName) versionRange)
-            <- getAllToolDependencies pkg_descr bi
-          , not $ isInternal pkg_descr buildTool ]
+    let requiredBuildTools = do
+          bi <- enabledBuildInfos pkg_descr enabled
+          -- First, we collect any tool dep that we know is external. This is,
+          -- in practice:
+          --
+          -- 1. `build-tools` entries on the whitelist
+          --
+          -- 2. `build-tool-depends` that aren't from the current package.
+          let externBuildToolDeps =
+                [ LegacyExeDependency (unUnqualComponentName eName) versionRange
+                | buildTool@(ExeDependency _ (Just eName) versionRange)
+                  <- getAllToolDependencies pkg_descr bi
+                , not $ isInternal pkg_descr buildTool ]
+          -- Second, we collect any build-tools entry we don't know how to
+          -- desugar. We'll never have no idea how to build them, so we just
+          -- hope they are already on the PATH.
+          let unknownBuildTools =
```

Comment:
I just realized that I may have communicated the situation here a little poorly. It is true that Setup was modified to pick up unknown build-tools from PATH, but in fact, all the testing code we have does not seem to actually exercise this case, as it is always assumed the build-tools refers to an internal dep. So a test here would be appreciated.